aicsimageio/aics_series.py

```diff
@@ -0,0 +1,367 @@
+
+log = logging.getLogger(__name__)
+
+###############################################################################
+
+
+class AICSSeries:
+    """
+    AICSSeries takes an ordered iterable of image data types (files / bytestreams) of consistent dimensions and forces
+    them to act like they are a single image. The data for each image in the series is lazy loaded as are the images
+    themselves. This is beneficial when the images stacked together may be too large to fit in memory.
+
+    Examples
+    --------
+    series = AICSSeries(["img1.tiff", "img2.tiff"], "T")
+    series[0, :, :, 3, :, :]  # returns a 4D array of TCYX
+
+    series = AICSImage("directory_full_of_tiffs/", "Z")
+    series[0, 0, 0, 3, :, :]  # returns a 2D array of YX
+
+    Notes
+    -----
+    * Because each image is lazy loaded, we validate on every image read. If the image has different dimensions or
+    sizes from the prior read image, an InvalidDimensionOrderingError or InconsitentDataShapeException is raised.
+    * This class may be a temporary stop gap until chunked image reading is added to the core `AICSImage` class and if
+    and when that occurs, this class will be depreciated.
+    * No metadata is checked for ensuring that the data in each image is consistent across them all. Example being,
+    channels in different indicies in different files across the series.
+    """
+
+    def __init__(
+        self,
+        images: Union[types.PathLike, Iterable[types.PathLike]],
+        series_dim: str
+    ):
+        """
+        Constructor for AICSSeries class intended for providing an interface for image reading of a list of file like
+        references.
+
+        Parameters
+        ----------
+        images: Union[types.PathLike, Iterable[types.PathLike]]
+            Either a pathlike value to a directory with images or the ordered iterable set of pathlike objects to read
+            from for the series.
+        series_dim: str
+            Which dimension to override with the iterable images.
+        """
+        # Check if provided a directory, if so, get the directory contents
+        if isinstance(images, (str, Path)):
+            # Expand and handle path
+            images = Path(images).expanduser().resolve(strict=True)
+
+            # Ensure that it is actually a directory and not a single file
+            if not images.is_dir():
+                raise NotADirectoryError(
+                    f"Provided a single file: {images}. Did you mean to use `AICSImage`?"
+                )
+
+            # Set images to the contents of the directory
+            images = sorted(images.iterdir())
+
+            # Drop any files that start with `.`
+            # TODO: Is this actually a valid operation to do for the user or way to specific?
+            # TODO: Different option, find the most common suffix and use it as the base?
+            predrop_count = len(images)
+            images = [img for img in images if img.name[0] != "."]
```

Comment:
I'm not sure about this maybe the processing above addresses this but what happens if for some strange reason the user chose to name their files 
[.img1.tiff, .img2.tiff, .img3.tiff] 
given that they are passing them as an explicit list shouldn't they be able to do so?

Reply:
I agree with Jamie that it doesn't make sense to load a series of mixed file format extensions. We could reject earlier if that scenario comes up.  I don't think this code needs to check for missing indices, since it is not making any other assumptions about naming conventions.  I also agree that a data type check is a good thing to raise exceptions on - files should match in shape and dtype.